storage: ignore non-intent locks in `intentInterleavingIter`

nvb opened the pull request:
Informs #100193.
Closes #109644.

This commit updates the intentInterleavingIter to ignore locks in the lock table keyspace with strengths other than lock.Intent (i.e. shared and exclusive locks). Future versions of the iterator may expose information to users about whether any non-intent locks were observed and, if so, which keys they were found on. For now, no such information is exposed.

This change is needed for replicated locks. Non-locking reads will not interact with these locks at all, so we want the intentInterleavingIter to ignore them. Meanwhile, locking reads and writes will use a different abstraction to peek into the lock table and look for conflicts.

<details>
  <summary>Benchmark delta without Locks</summary>
  
  ```
name                                                                     old time/op    new time/op    delta
IntentInterleavingIterNext/version=1/intentStride=100/keyLen=100-10         112ns ± 7%     111ns ± 4%    ~     (p=0.332 n=25+22)
IntentInterleavingIterNext/version=1/intentStride=1000000/keyLen=10-10     61.6ns ± 4%    61.1ns ± 2%    ~     (p=0.340 n=24+20)
IntentInterleavingIterNext/version=1/intentStride=1000000/keyLen=100-10    66.2ns ± 5%    66.1ns ± 4%    ~     (p=0.788 n=25+25)
IntentInterleavingIterNext/version=5/intentStride=1/keyLen=100-10           115ns ±10%     114ns ± 4%    ~     (p=0.367 n=25+25)
IntentInterleavingIterNext/version=5/intentStride=1000000/keyLen=10-10     58.9ns ± 4%    59.7ns ± 4%    ~     (p=0.051 n=21+25)
IntentInterleavingIterNext/version=5/intentStride=1000000/keyLen=100-10    63.7ns ± 3%    64.2ns ± 4%    ~     (p=0.100 n=24+25)
IntentInterleavingIterPrev/version=1/intentStride=100/keyLen=100-10         123ns ± 8%     124ns ± 4%    ~     (p=0.081 n=25+24)
IntentInterleavingIterNext/version=5/intentStride=100/keyLen=100-10         109ns ± 4%     110ns ± 4%  +0.81%  (p=0.002 n=24+24)
IntentInterleavingIterNext/version=5/intentStride=100/keyLen=10-10          100ns ± 4%     102ns ± 1%  +1.49%  (p=0.003 n=24+20)
IntentInterleavingIterPrev/version=5/intentStride=1/keyLen=10-10           84.9ns ± 4%    86.3ns ± 4%  +1.60%  (p=0.000 n=24+24)
IntentInterleavingIterPrev/version=1/intentStride=1000000/keyLen=100-10     122ns ± 6%     124ns ± 4%  +1.90%  (p=0.001 n=25+24)
IntentInterleavingIterPrev/version=5/intentStride=1000000/keyLen=100-10     123ns ± 4%     126ns ± 7%  +2.44%  (p=0.000 n=24+24)
IntentInterleavingIterNext/version=1/intentStride=100/keyLen=10-10          101ns ± 4%     104ns ± 5%  +2.46%  (p=0.000 n=25+24)
IntentInterleavingIterNext/version=5/intentStride=1/keyLen=10-10            103ns ± 6%     105ns ± 5%  +2.64%  (p=0.000 n=23+24)
IntentInterleavingIterPrev/version=5/intentStride=100/keyLen=100-10         122ns ± 3%     126ns ± 5%  +2.75%  (p=0.000 n=22+25)
IntentInterleavingIterPrev/version=5/intentStride=1/keyLen=100-10          92.0ns ± 2%    94.8ns ± 6%  +3.13%  (p=0.000 n=23+24)
IntentInterleavingIterPrev/version=1/intentStride=1000000/keyLen=10-10      112ns ± 4%     116ns ± 6%  +3.24%  (p=0.000 n=25+25)
IntentInterleavingIterPrev/version=1/intentStride=100/keyLen=10-10          112ns ± 5%     116ns ± 4%  +3.36%  (p=0.000 n=25+25)
IntentInterleavingIterPrev/version=5/intentStride=100/keyLen=10-10          112ns ± 3%     116ns ± 4%  +3.65%  (p=0.000 n=25+25)
IntentInterleavingIterNext/version=1/intentStride=1/keyLen=100-10           112ns ± 8%     117ns ± 1%  +3.83%  (p=0.000 n=25+20)
IntentInterleavingIterPrev/version=5/intentStride=1000000/keyLen=10-10      111ns ± 2%     116ns ± 3%  +4.02%  (p=0.000 n=21+24)
IntentInterleavingIterNext/version=1/intentStride=1/keyLen=10-10            103ns ± 4%     107ns ± 3%  +4.07%  (p=0.000 n=25+21)
IntentInterleavingIterPrev/version=1/intentStride=1/keyLen=100-10           105ns ± 3%     111ns ± 6%  +6.23%  (p=0.000 n=24+25)
IntentInterleavingIterPrev/version=1/intentStride=1/keyLen=10-10           94.4ns ± 3%   100.5ns ± 5%  +6.48%  (p=0.000 n=23+25)

name                                                                     old alloc/op   new alloc/op   delta
IntentInterleavingIterNext/version=1/intentStride=1/keyLen=10-10            0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=1/intentStride=1/keyLen=100-10           0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=1/intentStride=100/keyLen=10-10          0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=1/intentStride=100/keyLen=100-10         0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=1/intentStride=1000000/keyLen=10-10      0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=1/intentStride=1000000/keyLen=100-10     0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=1/keyLen=10-10            0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=1/keyLen=100-10           0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=100/keyLen=10-10          0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=100/keyLen=100-10         0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=1000000/keyLen=10-10      0.00B          0.00B         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=1000000/keyLen=100-10     0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=1/keyLen=10-10            0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=1/keyLen=100-10           0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=100/keyLen=10-10          0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=100/keyLen=100-10         0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=1000000/keyLen=10-10      0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=1000000/keyLen=100-10     0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=1/keyLen=10-10            0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=1/keyLen=100-10           0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=100/keyLen=10-10          0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=100/keyLen=100-10         0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=1000000/keyLen=10-10      0.00B          0.00B         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=1000000/keyLen=100-10     0.00B          0.00B         ~     (all equal)

name                                                                     old allocs/op  new allocs/op  delta
IntentInterleavingIterNext/version=1/intentStride=1/keyLen=10-10             0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=1/intentStride=1/keyLen=100-10            0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=1/intentStride=100/keyLen=10-10           0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=1/intentStride=100/keyLen=100-10          0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=1/intentStride=1000000/keyLen=10-10       0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=1/intentStride=1000000/keyLen=100-10      0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=1/keyLen=10-10             0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=1/keyLen=100-10            0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=100/keyLen=10-10           0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=100/keyLen=100-10          0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=1000000/keyLen=10-10       0.00           0.00         ~     (all equal)
IntentInterleavingIterNext/version=5/intentStride=1000000/keyLen=100-10      0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=1/keyLen=10-10             0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=1/keyLen=100-10            0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=100/keyLen=10-10           0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=100/keyLen=100-10          0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=1000000/keyLen=10-10       0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=1/intentStride=1000000/keyLen=100-10      0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=1/keyLen=10-10             0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=1/keyLen=100-10            0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=100/keyLen=10-10           0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=100/keyLen=100-10          0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=1000000/keyLen=10-10       0.00           0.00         ~     (all equal)
IntentInterleavingIterPrev/version=5/intentStride=1000000/keyLen=100-10      0.00           0.00         ~     (all equal)
  ```
</details>

<details>
  <summary>Benchmarks with Locks</summary>
  
  ```
name                                                                                        time/op
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1/keyLen=10-10                149ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1/keyLen=100-10               161ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=100/keyLen=10-10              105ns ± 0%
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=100/keyLen=100-10             113ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1000000/keyLen=10-10          104ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1000000/keyLen=100-10         112ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1/keyLen=10-10              173ns ± 2%
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1/keyLen=100-10             187ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=100/keyLen=10-10           94.3ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=100/keyLen=100-10           102ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1000000/keyLen=10-10       93.0ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1000000/keyLen=100-10       100ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1/keyLen=10-10          172ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1/keyLen=100-10         188ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=100/keyLen=10-10       93.6ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=100/keyLen=100-10       101ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1000000/keyLen=10-10   54.1ns ± 1%
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1000000/keyLen=100-10  58.8ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1/keyLen=10-10                112ns ± 0%
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1/keyLen=100-10               121ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=100/keyLen=10-10             96.6ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=100/keyLen=100-10             105ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1000000/keyLen=10-10         96.6ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1000000/keyLen=100-10         104ns ± 0%
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1/keyLen=10-10              108ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1/keyLen=100-10             118ns ± 0%
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=100/keyLen=10-10           91.5ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=100/keyLen=100-10          98.6ns ± 0%
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1000000/keyLen=10-10       91.2ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1000000/keyLen=100-10      98.5ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1/keyLen=10-10          108ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1/keyLen=100-10         118ns ± 0%
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=100/keyLen=10-10       91.6ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=100/keyLen=100-10      98.9ns ± 1%
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1000000/keyLen=10-10   51.6ns ± 0%
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1000000/keyLen=100-10  56.5ns ± 2%
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1/keyLen=10-10                149ns ± 1%
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1/keyLen=100-10               159ns ± 1%
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=100/keyLen=10-10             96.7ns ± 1%
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=100/keyLen=100-10             105ns ± 5%
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1000000/keyLen=10-10         94.2ns ± 0%
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1000000/keyLen=100-10         103ns ± 0%
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1/keyLen=10-10              190ns ± 0%
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1/keyLen=100-10             207ns ± 0%
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=100/keyLen=10-10            103ns ± 1%
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=100/keyLen=100-10           111ns ± 1%
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1000000/keyLen=10-10        102ns ± 1%
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1000000/keyLen=100-10       110ns ± 1%
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1/keyLen=10-10          190ns ± 0%
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1/keyLen=100-10         207ns ± 0%
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=100/keyLen=10-10        102ns ± 1%
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=100/keyLen=100-10       111ns ± 1%
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1000000/keyLen=10-10    101ns ± 0%
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1000000/keyLen=100-10   110ns ± 1%
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1/keyLen=10-10               94.7ns ± 0%
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1/keyLen=100-10               103ns ± 1%
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=100/keyLen=10-10             77.8ns ± 1%
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=100/keyLen=100-10            86.4ns ± 1%
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1000000/keyLen=10-10         78.5ns ± 0%
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1000000/keyLen=100-10        86.3ns ± 0%
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1/keyLen=10-10              123ns ± 0%
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1/keyLen=100-10             135ns ± 1%
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=100/keyLen=10-10            103ns ± 2%
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=100/keyLen=100-10           111ns ± 0%
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1000000/keyLen=10-10        102ns ± 1%
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1000000/keyLen=100-10       111ns ± 1%
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1/keyLen=10-10          121ns ± 0%
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1/keyLen=100-10         133ns ± 2%
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=100/keyLen=10-10        102ns ± 0%
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=100/keyLen=100-10       112ns ± 1%
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1000000/keyLen=10-10    102ns ± 0%
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1000000/keyLen=100-10   111ns ± 0%

name                                                                                        alloc/op
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1/keyLen=10-10                0.00B
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1/keyLen=100-10               0.00B
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=100/keyLen=10-10              0.00B
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=100/keyLen=100-10             0.00B
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1000000/keyLen=10-10          0.00B
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1000000/keyLen=100-10         0.00B
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1/keyLen=10-10              0.00B
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1/keyLen=100-10             0.00B
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=100/keyLen=10-10            0.00B
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=100/keyLen=100-10           0.00B
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1000000/keyLen=10-10        0.00B
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1000000/keyLen=100-10       0.00B
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1/keyLen=10-10          0.00B
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1/keyLen=100-10         0.00B
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=100/keyLen=10-10        0.00B
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=100/keyLen=100-10       0.00B
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1000000/keyLen=10-10    0.00B
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1000000/keyLen=100-10   0.00B
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1/keyLen=10-10                0.00B
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1/keyLen=100-10               0.00B
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=100/keyLen=10-10              0.00B
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=100/keyLen=100-10             0.00B
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1000000/keyLen=10-10          0.00B
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1000000/keyLen=100-10         0.00B
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1/keyLen=10-10              0.00B
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1/keyLen=100-10             0.00B
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=100/keyLen=10-10            0.00B
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=100/keyLen=100-10           0.00B
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1000000/keyLen=10-10        0.00B
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1000000/keyLen=100-10       0.00B
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1/keyLen=10-10          0.00B
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1/keyLen=100-10         0.00B
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=100/keyLen=10-10        0.00B
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=100/keyLen=100-10       0.00B
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1000000/keyLen=10-10    0.00B
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1000000/keyLen=100-10   0.00B
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1/keyLen=10-10                0.00B
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1/keyLen=100-10               0.00B
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=100/keyLen=10-10              0.00B
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=100/keyLen=100-10             0.00B
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1000000/keyLen=10-10          0.00B
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1000000/keyLen=100-10         0.00B
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1/keyLen=10-10              0.00B
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1/keyLen=100-10             0.00B
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=100/keyLen=10-10            0.00B
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=100/keyLen=100-10           0.00B
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1000000/keyLen=10-10        0.00B
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1000000/keyLen=100-10       0.00B
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1/keyLen=10-10          0.00B
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1/keyLen=100-10         0.00B
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=100/keyLen=10-10        0.00B
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=100/keyLen=100-10       0.00B
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1000000/keyLen=10-10    0.00B
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1000000/keyLen=100-10   0.00B
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1/keyLen=10-10                0.00B
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1/keyLen=100-10               0.00B
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=100/keyLen=10-10              0.00B
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=100/keyLen=100-10             0.00B
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1000000/keyLen=10-10          0.00B
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1000000/keyLen=100-10         0.00B
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1/keyLen=10-10              0.00B
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1/keyLen=100-10             0.00B
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=100/keyLen=10-10            0.00B
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=100/keyLen=100-10           0.00B
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1000000/keyLen=10-10        0.00B
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1000000/keyLen=100-10       0.00B
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1/keyLen=10-10          0.00B
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1/keyLen=100-10         0.00B
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=100/keyLen=10-10        0.00B
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=100/keyLen=100-10       0.00B
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1000000/keyLen=10-10    0.00B
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1000000/keyLen=100-10   0.00B

name                                                                                        allocs/op
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1/keyLen=10-10                 0.00
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1/keyLen=100-10                0.00
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=100/keyLen=10-10               0.00
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=100/keyLen=100-10              0.00
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1000000/keyLen=10-10           0.00
IntentInterleavingIterNext/version=1/intentStride=1/lockStride=1000000/keyLen=100-10          0.00
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1/keyLen=10-10               0.00
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1/keyLen=100-10              0.00
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=100/keyLen=10-10             0.00
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=100/keyLen=100-10            0.00
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1000000/keyLen=10-10         0.00
IntentInterleavingIterNext/version=1/intentStride=100/lockStride=1000000/keyLen=100-10        0.00
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1/keyLen=10-10           0.00
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1/keyLen=100-10          0.00
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=100/keyLen=10-10         0.00
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=100/keyLen=100-10        0.00
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1000000/keyLen=10-10     0.00
IntentInterleavingIterNext/version=1/intentStride=1000000/lockStride=1000000/keyLen=100-10    0.00
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1/keyLen=10-10                 0.00
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1/keyLen=100-10                0.00
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=100/keyLen=10-10               0.00
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=100/keyLen=100-10              0.00
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1000000/keyLen=10-10           0.00
IntentInterleavingIterNext/version=5/intentStride=1/lockStride=1000000/keyLen=100-10          0.00
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1/keyLen=10-10               0.00
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1/keyLen=100-10              0.00
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=100/keyLen=10-10             0.00
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=100/keyLen=100-10            0.00
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1000000/keyLen=10-10         0.00
IntentInterleavingIterNext/version=5/intentStride=100/lockStride=1000000/keyLen=100-10        0.00
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1/keyLen=10-10           0.00
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1/keyLen=100-10          0.00
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=100/keyLen=10-10         0.00
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=100/keyLen=100-10        0.00
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1000000/keyLen=10-10     0.00
IntentInterleavingIterNext/version=5/intentStride=1000000/lockStride=1000000/keyLen=100-10    0.00
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1/keyLen=10-10                 0.00
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1/keyLen=100-10                0.00
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=100/keyLen=10-10               0.00
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=100/keyLen=100-10              0.00
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1000000/keyLen=10-10           0.00
IntentInterleavingIterPrev/version=1/intentStride=1/lockStride=1000000/keyLen=100-10          0.00
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1/keyLen=10-10               0.00
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1/keyLen=100-10              0.00
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=100/keyLen=10-10             0.00
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=100/keyLen=100-10            0.00
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1000000/keyLen=10-10         0.00
IntentInterleavingIterPrev/version=1/intentStride=100/lockStride=1000000/keyLen=100-10        0.00
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1/keyLen=10-10           0.00
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1/keyLen=100-10          0.00
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=100/keyLen=10-10         0.00
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=100/keyLen=100-10        0.00
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1000000/keyLen=10-10     0.00
IntentInterleavingIterPrev/version=1/intentStride=1000000/lockStride=1000000/keyLen=100-10    0.00
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1/keyLen=10-10                 0.00
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1/keyLen=100-10                0.00
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=100/keyLen=10-10               0.00
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=100/keyLen=100-10              0.00
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1000000/keyLen=10-10           0.00
IntentInterleavingIterPrev/version=5/intentStride=1/lockStride=1000000/keyLen=100-10          0.00
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1/keyLen=10-10               0.00
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1/keyLen=100-10              0.00
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=100/keyLen=10-10             0.00
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=100/keyLen=100-10            0.00
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1000000/keyLen=10-10         0.00
IntentInterleavingIterPrev/version=5/intentStride=100/lockStride=1000000/keyLen=100-10        0.00
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1/keyLen=10-10           0.00
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1/keyLen=100-10          0.00
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=100/keyLen=10-10         0.00
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=100/keyLen=100-10        0.00
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1000000/keyLen=10-10     0.00
IntentInterleavingIterPrev/version=5/intentStride=1000000/lockStride=1000000/keyLen=100-10    0.00
  ```
</details>

Release note: None

nvb committed:
storage: remove !separated variants from IntentInterleavingIter benchmarks

All intents are now separated.

Epic: None
Release note: None

nvb committed:
storage: ignore non-intent locks in intentInterleavingIter

Informs #100193.

This commit updates the intentInterleavingIter to ignore locks
in the lock table keyspace with strengths other than lock.Intent
(i.e. shared and exclusive locks). Future versions of the
iterator may expose information to users about whether any
non-intent locks were observed and, if so, which keys they were
found on. For now, no such information is exposed.

This change is needed for replicated locks. Non-locking reads
will not interact with these locks at all, so we want the
intentInterleavingIter to ignore them. Meanwhile, locking reads
and writes will use a different abstraction to peek into the
lock table and look for conflicts.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/109570)
<!-- Reviewable:end -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo comments

> Reviewed 1 of 1 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-:-Ncsz21aEDCwrVHRRUab:b-913mr5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[`pkg/storage/intent_interleaving_iter.go` line 716 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Ncspj_-DBNGfb2-Ivna:-Ncspj_-DBNGfb2-Ivnb:b-i7wluq) ([raw file](https://github.com/cockroachdb/cockroach/blob/43235931871d708245b498dff8c0837c89c96bcd/pkg/storage/intent_interleaving_iter.go#L716)):*
> ```Go
> // seeks intentIter to the first intent key equal to (if i.dir > 0) or after
> // seekKey and decodes it into i.intentKey.
> func (i *intentInterleavingIter) seekIntentIterAndDecodeLockKey(
> ```

nit: seekIntentIterAndDecodeLockKey seeks ...
___
*[`pkg/storage/intent_interleaving_iter.go` line 719 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-NcsrDLA866DvxTM10Sg:-NcsrDLA866DvxTM10Sh:b5ewbp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/43235931871d708245b498dff8c0837c89c96bcd/pkg/storage/intent_interleaving_iter.go#L719)):*
> ```Go
> 	seekKey EngineKey, limitKey roachpb.Key,
> ) (iterState pebble.IterValidityState, err error) {
> 	if i.dir < 0 {
> ```

I see we do this elsewhere, so feel free to disregard to keep things uniform, but would this logic be better expressed as:

```go
switch i.dir {  
case 1:  
 iterState, err = i.intentIter.SeekEngineKeyGEWithLimit(seekKey, limitKey)  
case -1:  
 iterState, err = i.intentIter.SeekEngineKeyLTWithLimit(seekKey, limitKey)  
default:  
 return iterState, errors.AssertionFailedf(" ... ")  
}
```
___
*[`pkg/storage/intent_interleaving_iter.go` line 731 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-NcspoV_3CUvaXfIJcjl:-NcspoV_3CUvaXfIJcjm:b1rvwf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/43235931871d708245b498dff8c0837c89c96bcd/pkg/storage/intent_interleaving_iter.go#L731)):*
> ```Go
> 
> // steps intentIter to the next intent key and decodes it into i.intentKey.
> func (i *intentInterleavingIter) stepIntentIterAndDecodeLockKey(
> ```

nit: stepIntentIterAndDecodeLockKey steps ...
___
*[`pkg/storage/intent_interleaving_iter_test.go` line 588 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Ncsy5Xb1q1OC8wBKbQk:-Ncsy5Xb1q1OC8wBKbQl:b4ftnor) ([raw file](https://github.com/cockroachdb/cockroach/blob/43235931871d708245b498dff8c0837c89c96bcd/pkg/storage/intent_interleaving_iter_test.go#L588)):*
> ```Go
> 			// Lock.
> 			str := lock.Shared
> 			if rng.Int31n(2) == 0 {
> ```

If you take the suggestion about generating more than 1 shared locks, we might want to pull this outside the loop. 
___
*[`pkg/storage/intent_interleaving_iter_test.go` line 598 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-NcswLIHBdJWEdRf0utT:-NcswLIHBdJWEdRf0utU:bq2nnjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/43235931871d708245b498dff8c0837c89c96bcd/pkg/storage/intent_interleaving_iter_test.go#L598)):*
> ```Go
> 			lockLTKey := LockTableKey{Key: key, Strength: str, TxnUUID: txnUUID}
> 			lkv = append(lkv, lockKeyValue{
> 				key: lockLTKey, val: lockVal, liveKey: hasLock && i == 0})
> ```

Given multiple shared locks can be compatible with each other, if we're in the `str == lock.Shared` case, would it be interesting to generate more than 1 shared locks here? 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109570)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/storage/intent_interleaving_iter.go` line 716 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Ncspj_-DBNGfb2-Ivna:-Nd0mU3s7g_7hAWIIfIZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/43235931871d708245b498dff8c0837c89c96bcd/pkg/storage/intent_interleaving_iter.go#L716)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: seekIntentIterAndDecodeLockKey seeks ...
</blockquote></details>

Done.
___
*[`pkg/storage/intent_interleaving_iter.go` line 719 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-NcsrDLA866DvxTM10Sg:-Nd0mZwb2AUB79UT0-5N:b5calia) ([raw file](https://github.com/cockroachdb/cockroach/blob/43235931871d708245b498dff8c0837c89c96bcd/pkg/storage/intent_interleaving_iter.go#L719)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I see we do this elsewhere, so feel free to disregard to keep things uniform, but would this logic be better expressed as:

```go
switch i.dir {  
case 1:  
 iterState, err = i.intentIter.SeekEngineKeyGEWithLimit(seekKey, limitKey)  
case -1:  
 iterState, err = i.intentIter.SeekEngineKeyLTWithLimit(seekKey, limitKey)  
default:  
 return iterState, errors.AssertionFailedf(" ... ")  
}
```
</blockquote></details>

Yeah, `i.dir < 0` and `i.dir > 0` seem to be the preferred way to use this field in the file, so I'll follow the same pattern here.
___
*[`pkg/storage/intent_interleaving_iter.go` line 731 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-NcspoV_3CUvaXfIJcjl:-Nd0mVenC8EzMFEy5nXk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/43235931871d708245b498dff8c0837c89c96bcd/pkg/storage/intent_interleaving_iter.go#L731)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: stepIntentIterAndDecodeLockKey steps ...
</blockquote></details>

Done.
___
*[`pkg/storage/intent_interleaving_iter_test.go` line 588 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Ncsy5Xb1q1OC8wBKbQk:-Nd0q9gXDHFswjLs2o8o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/43235931871d708245b498dff8c0837c89c96bcd/pkg/storage/intent_interleaving_iter_test.go#L588)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

If you take the suggestion about generating more than 1 shared locks, we might want to pull this outside the loop.
</blockquote></details>

Done.
___
*[`pkg/storage/intent_interleaving_iter_test.go` line 598 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-NcswLIHBdJWEdRf0utT:-Nd0qAQ89cv5mJKqBST3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/43235931871d708245b498dff8c0837c89c96bcd/pkg/storage/intent_interleaving_iter_test.go#L598)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Given multiple shared locks can be compatible with each other, if we're in the `str == lock.Shared` case, would it be interesting to generate more than 1 shared locks here?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109570)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/storage/intent_interleaving_iter.go` line 586 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Nd1BW0o30VQX4Qbw6hZ:-Nd1BW0o30VQX4Qbw6h_:b-8gk05b) ([raw file](https://github.com/cockroachdb/cockroach/blob/339bf7a67f93d8a7e7c88623b3824ff81e441494/pkg/storage/intent_interleaving_iter.go#L586)):*
> ```Go
> 			limitKey = i.makeUpperLimitKey()
> 		}
> 		if _, err := i.seekIntentIterAndDecodeLockKey(EngineKey{Key: intentSeekKey}, limitKey); err != nil {
> ```

Given that the strength is the first byte in the version, I think we can do better and seek to the versions that begin with the intent strength. Similarly for the SeekLT. That should eliminate the need to iterate past non-intent locks in these seek cases.

For `intentInterleavingIter.Prev`, since byte 3 is going to sort before the others, stepping back will first encounter the other locks of the preceding `roachpb.Key`. For `intentIntentInterleaving.Next`, stepping forward will first encounter the other locks of the same `roachpb.Key`. So for both these cases (where we call `NextEngineKeyWithLimit` or `PrevEngineKeyWithLimit`) we do need iteration. The one concern I have with introducing a loop where we didn't have one before (and I haven't thought through this properly), is the few cases of the form
```
var limitKey roachpb.Key  
if i.iterValid && !i.prefix {  
 limitKey = i.makeUpperLimitKey()  
}  
iterState, err := i.intentIter.NextEngineKeyWithLimit(limitKey)
```
where we are not in prefix iteration AND we don't produce a `limitKey`. We won't stop until we hit the iterator bounds if there are no intents but lot of other locks. These cases need some examination, since continuing until the iterator bounds is not good.
___
*[`pkg/storage/intent_interleaving_iter.go` line 778 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Nd16zSkEsg4FS_LBZx6:-Nd16zSkEsg4FS_LBZx7:b-yxvt8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/339bf7a67f93d8a7e7c88623b3824ff81e441494/pkg/storage/intent_interleaving_iter.go#L778)):*
> ```Go
> 		return false, err
> 	}
> 	lockTableKey, err := engineKey.ToLockTableKey()
> ```

This could be optimized if we often expect to find non-intent locks, since there is no need to do all the work in `ToLockTableKey`. A TODO is fine.


<!-- Sent from Reviewable.io -->

nvb committed:
[DNM] storage: add dynamic iter/seek adjustment to intentInterleavingIter

This commit configures the intentInterleavingIter to dynamically adjust
the number of iterations that it will try across the locks on single
user key while searching for intents in the lock table before it does a
SeekGE/SeekLT to skip to the locks on the next/previous user key.

This places a bound on the amount of work that non-locking iteration
will perform for keys with many (shared) locks. As a result, scan speed
will no longer be proportional to the total number of locks on the keys
that it scans over. Scan speed will still be proportional to the number
of keys that have locks, but we expect that keys with locks will also
have values, so this per-locked-key cost is acceptable.

Don't merge, this needs more testing.

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109570)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/storage/intent_interleaving_iter.go` line 586 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Nd1BW0o30VQX4Qbw6hZ:-NdczoP9C76vVS-3Nmbw:bje60h5) ([raw file](https://github.com/cockroachdb/cockroach/blob/339bf7a67f93d8a7e7c88623b3824ff81e441494/pkg/storage/intent_interleaving_iter.go#L586)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Given that the strength is the first byte in the version, I think we can do better and seek to the versions that begin with the intent strength. Similarly for the SeekLT. That should eliminate the need to iterate past non-intent locks in these seek cases.

For `intentInterleavingIter.Prev`, since byte 3 is going to sort before the others, stepping back will first encounter the other locks of the preceding `roachpb.Key`. For `intentIntentInterleaving.Next`, stepping forward will first encounter the other locks of the same `roachpb.Key`. So for both these cases (where we call `NextEngineKeyWithLimit` or `PrevEngineKeyWithLimit`) we do need iteration. The one concern I have with introducing a loop where we didn't have one before (and I haven't thought through this properly), is the few cases of the form
```
var limitKey roachpb.Key  
if i.iterValid && !i.prefix {  
 limitKey = i.makeUpperLimitKey()  
}  
iterState, err := i.intentIter.NextEngineKeyWithLimit(limitKey)
```
where we are not in prefix iteration AND we don't produce a `limitKey`. We won't stop until we hit the iterator bounds if there are no intents but lot of other locks. These cases need some examination, since continuing until the iterator bounds is not good.
</blockquote></details>

@sumeerbhola I added a third commit here which configures the `intentInterleavingIter` to dynamically adjust the number of iterations that it will try across the locks on single user key before it does a SeekGE/SeekLT to skip to the locks on the next/previous user key. The commit also short-circuits without any version iteration for prefix iterators.

I believe this is where we landed after our conversation last week, though I wasn't sure if you still thought the `limitKey` could be integrated into this logic. Before I add a set of tests for the behavior, do you mind verifying that it's what you expect and that we should proceed with the approach? 
___
*[`pkg/storage/intent_interleaving_iter.go` line 778 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Nd16zSkEsg4FS_LBZx6:-Ndczmq9-DPZvma-ByAa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/339bf7a67f93d8a7e7c88623b3824ff81e441494/pkg/storage/intent_interleaving_iter.go#L778)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This could be optimized if we often expect to find non-intent locks, since there is no need to do all the work in `ToLockTableKey`. A TODO is fine.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109570)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/storage/intent_interleaving_iter.go` line 228 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Ne3l5z0EonbbK3IXvTo:-Ne3l5z0EonbbK3IXvTp:b-cejhav) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cd9334a67b1d96139adfac49cc84b96e0407d9c/pkg/storage/intent_interleaving_iter.go#L228)):*
> ```Go
> 	// adjustment algorithm that is used by pebbleMVCCScanner when scanning over
> 	// mvcc versions for a key.
> 	maxIntentItersBeforeSeek int
> ```

I don't see where we are updating this value, for the "dynamic adjustment"
___
*[`pkg/storage/intent_interleaving_iter.go` line 760 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Ne3lpX9D0BaLan27bs4:-Ne3lpX9D0BaLan27bs5:b6gbbi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cd9334a67b1d96139adfac49cc84b96e0407d9c/pkg/storage/intent_interleaving_iter.go#L760)):*
> ```Go
> 	// Otherwise, iterate over the locks on each key until we find an intent.
> 	itersBeforeSeek := i.maxIntentItersBeforeSeek / 2
> 	if itersBeforeSeek == 0 && i.intentIterState != pebble.IterValid {
> ```

we need a valid iterator to do a seek, so makes sense that we are stepping when `i.intentIterState != pebble.IterValid` but why is this a conjunction with `itersBeforeSeek == 0`?
___
*[`pkg/storage/intent_interleaving_iter.go` line 823 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/109570#-Ne3mchSELEQQ8E3z8u7:-Ne3mchSELEQQ8E3z8u8:b-q9gg1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cd9334a67b1d96139adfac49cc84b96e0407d9c/pkg/storage/intent_interleaving_iter.go#L823)):*
> ```Go
> 			}
> 		} else {
> 			i.intentKeyBuf = append(i.intentKeyBuf[:0], i.intentKey...)
> ```

this will only happen in the first iteration, yes?


<!-- Sent from Reviewable.io -->
